cypress/plugins/synapsedocker/index.ts

```diff
@@ -0,0 +1,169 @@
+    configDir: string;
+    registrationSecret: string;
+}
+
+export interface SynapseInstance extends SynapseConfig {
+    synapseId: string;
+    port: number;
+}
+
+const synapses = new Map<string, SynapseInstance>();
+
+function randB64Bytes(numBytes: number): string {
+    return crypto.randomBytes(numBytes).toString("base64").replace(/=*$/, "");
+}
+
+async function cfgDirFromTemplate(template: string): Promise<SynapseConfig> {
+    const templateDir = path.join(__dirname, "templates", template);
+
+    const stats = await fse.stat(templateDir);
+    if (!stats || !stats.isDirectory) {
+        throw new Error(`No such template: ${template}`);
+    }
+    const tempDir = await fse.mkdtemp(path.join(os.tmpdir(), 'react-sdk-synapsedocker-'));
+
+    // change permissions on the temp directory so the docker container can see its contents
+    await fse.chmod(tempDir, 0o777);
+
+    // copy the contents of the template dir, omitting homeserver.yaml as we'll template that
+    console.log(`Copy ${templateDir} -> ${tempDir}`);
+    await fse.copy(templateDir, tempDir, { filter: f => path.basename(f) !== 'homeserver.yaml' });
+
+    const registrationSecret = randB64Bytes(16);
+    const macaroonSecret = randB64Bytes(16);
+    const formSecret = randB64Bytes(16);
+
+    // now copy homeserver.yaml, applying sustitutions
+    console.log(`Gen ${path.join(templateDir, "homeserver.yaml")}`);
+    let hsYaml = await fse.readFile(path.join(templateDir, "homeserver.yaml"), "utf8");
+    hsYaml = hsYaml.replace(/{{REGISTRATION_SECRET}}/g, registrationSecret);
+    hsYaml = hsYaml.replace(/{{MACAROON_SECRET_KEY}}/g, macaroonSecret);
+    hsYaml = hsYaml.replace(/{{FORM_SECRET}}/g, formSecret);
+    await fse.writeFile(path.join(tempDir, "homeserver.yaml"), hsYaml);
+
+    // now generate a signing key (we could use synapse's config generation for
+    // this, or we could just do this...)
+    // NB. This assumes the homeserver.yaml specifies the key in this location
+    const signingKey = randB64Bytes(32);
+    console.log(`Gen ${path.join(templateDir, "localhost.signing.key")}`);
+    await fse.writeFile(path.join(tempDir, "localhost.signing.key"), `ed25519 x ${signingKey}`);
+
+    return {
+        configDir: tempDir,
+        registrationSecret,
+    };
+}
+
+/**
+ * @type {Cypress.PluginConfig}
+ */
+// eslint-disable-next-line no-unused-vars
+export function synapseDocker(on, config) {
+    on("task", {
+        // Start a synapse instance: the template must be the name of
+        // one of the templates in the cypress/plugins/synapsedocker/templates
+        // directory
+        async synapseStart(template: string): Promise<SynapseInstance> {
+            const synCfg = await cfgDirFromTemplate(template);
+
+            console.log(`Starting synapse with config dir ${synCfg.configDir}...`);
+
+            const synapseId = await new Promise<string>((resolve, reject) => {
+                childProcess.execFile('docker', [
```

Comment:
can we also give this a `--name` and set `--rm` so it cleans itself up upon exiting?

Reply:
Have given them a name. We collect logs at the end so if we use --rm then we don't have a chance to get the logs out. I have also made it tidy up any remaining synapses after each spec run to try to avoid containers sitting around if people forget to stop them in the tests.